course3/homework-linked-list/linked_list.h

```diff
@@ -8,29 +8,66 @@ class LinkedList
 {
 public:
 
-    LinkedList();
+    LinkedList(): head(nullptr), tail(nullptr), length(0) { }
 
-    const T* begin() const;
-    T* begin();
-    const T* end() const;
-    T* end();
+    const T* begin() const {
+        return this->head ? this->head->data : nullptr;
+    }
 
-    std::size_t size() const;
-    bool empty() const;
+    T* begin() {
+        return this->head ? this->head->data : nullptr;
+    }
 
-    void push_front(const T&);
-    void pop_front();
+    const T* end() const {
+        return this->tail ? this->tail->data : nullptr;
+    }
 
-    const T& front() const;
-    T& front();
+    T* end() {
+        return this->tail ? this->tail->data : nullptr;
+    }
+
+    std::size_t size() const {
+        return this->length;
+    }
+
+    bool empty() const {
+        return this->length == 0;
+    }
+
+    void push_front(const T& input) {
+        auto n = Node(input);
+        n.next = this->head;
+        this->head = &n;
+        this->length++;
+    }
+
+    void pop_front() {
+        if (this->empty()) {
+            throw std::out_of_range("The list is empty");
+        }
+        auto topop = this->head;
+        this->head = this->head->next;
+        delete topop;
```

Comment:
Don't use explicit 'delete' or explicit 'new', they are breeding ground for bugs!